onnxruntime/test/shared_lib/custom_op_utils.cc

```diff
@@ -206,3 +207,18 @@ void SliceCustomOpKernel::Compute(OrtKernelContext* context) {
       ORT_THROW("Unsupported input type");
   }
 }
+
+void SimpleMultiThreadingCopyKernel::Compute(OrtKernelContext* context) {
+  using OrtThreadPool = onnxruntime::concurrency::ThreadPool*;
+  using ConstOrtThreadPool = const onnxruntime::concurrency::ThreadPool*;
+  auto threadpool = reinterpret_cast<ConstOrtThreadPool>(ort_.KernelContext_GetThreadPool(context));
+  const OrtValue* input = ort_.KernelContext_GetInput(context, 0);
+  OrtTensorDimensions dimensions(ort_, input);
+  OrtValue* output = ort_.KernelContext_GetOutput(context, 0, dimensions.data(), dimensions.size());
+  auto N = 4 * std::accumulate(dimensions.begin(), dimensions.end(), 1LL, std::multiplies<int64_t>());
+  const char* input_data = ort_.GetTensorData<char>(input);
+  char* output_data = ort_.GetTensorMutableData<char>(output);
+  onnxruntime::concurrency::ThreadPool::TrySimpleParallelFor(const_cast<OrtThreadPool>(threadpool), static_cast<std::ptrdiff_t>(N), [&](std::ptrdiff_t i) {
```

Comment:
onnxruntime::concurrency::ThreadPool is not ABI. It may have different layout across DLLs. So here when you cast the void* back it may fail.

Reply:
Similarly, what if we replace onnxruntime::concurrency::ThreadPool with std::string. Can we return a std::string in such a way? No. That's why our C API is so complicated. 

In the C API, you can return a pointer to an internal data structure only if the pointer will be passed back in the same way and the user will only use it as void*.